Save application

SSPJ opened the pull request:
## 🗣 Description ##

This PR adds all the plumbing needed to save and restore progress on domain applications.

## 💭 Motivation and context ##

This is a very big PR with a number of outstanding items and caveats.

- [ ] Add in progress and submitted examples of applications to our data fixtures
    - [ ] On that note, it would be nice to have our fixtures autoload whenever the docker containers start up
- [ ] Complete the unit tests for restore; they are currently marked with `@skip`
- [ ] A messaging system needs to be created and tested; currently the user has no way to know that pressing "save" did anything at all
- [ ]  There is an edge case where a user can select "Federal", click next, select an option ("Judicial", for example), click save or next, then go back and select, say, "State", and click next. Their database record will now say that they are of type "State" and of federal type "Judicial". This seems like a good candidate for `wontfix`.
- [ ] Other contacts, organization website, and alternative domains currently have no mechanism to add more than one
- [ ] Lastly, the combo boxes don't work as expected; the correct value is being "selected" in the underlying `<select>` element but does not show up visually on the page. I am in favor of switching to either a simple select or text input due to [other accessibility issues](https://github.com/uswds/uswds-site/issues/1898) in this component, but I'd like to get your opinion, @igorkorenfeld
- [ ]  Have `Domain` inherit from `Website` so that they can share all code related to the domain name string

Closes #260

## 🧪 Testing Instructions ##

Manual testing of save is quite easy:
1. Go to http://localhost:8080/register to begin a new application
2. Open the admin panel  http://localhost:8080/admin/registrar/domainapplication/ in another tab
3. On each page, fill in the fields and click save
4. Verify that the page refreshes and that the info you saved is now in the database model

Manual testing of restore is a bit fiddly. You'll need to create an application, fill in most fields, save it (but don't finish it!), then delete your cookies, then visit http://localhost:8080/application/1/edit (assuming the domain application primary key is `1`).

neilmb commented on src/registrar/models/website.py:
```diff
@@ -14,5 +16,35 @@ class Website(models.Model):
         help_text="",
     )
 
+    @property
```
**question (non-blocking)**: It looks like this code is close to duplicated with `domain`. I'm assuming that the abstraction necessary to keep from duplicating wasn't worth it in this case.

neilmb commented on src/registrar/tests/test_views.py:
```diff
@@ -378,3 +548,124 @@ def test_application_form_conditional_elections(self):
+        url = reverse("application_step", kwargs={"step": "organization_type"})
+        response = self.client.get(url, follow=True)
+        self.assertContains(response, "<input>")
+        # choices = response.context['wizard']['form']['organization_type'].subwidgets
```
**nitpick**: If we are going to comment out this much code, could we leave a TODO at the beginning of the block here about why it is commented out?

SSPJ replied:
Yup, can do. Also happy to finish these tests on this PR, if you'd like... but as you noted, the restore functionality is still somewhat tbd, so I would rather wait until the design for that is finalized.

neilmb reviewed (approved):
LGTM. So glad to see all of this. This is way more than just adding a save button, it sets us up for being able to edit domain applications as well, even though the interface might not be exactly what we want in the end.

Nothing but nitpicks, feel free to merge as-is.

igorkorenfeld reviewed (commented):
Thanks on getting this working! 

Unfortunately I am not able to get through the form, I get an error on the `Organization Website` (pasted the debugging info below). I got the error both when pressing "Save" with a blank field, when entering a string and clicking the "Next" button, and with a blank field and clicking the "Next" button

Because I never get to the part where I enter my requested dotgov domain, I also can't get back into it – we need some sort of placeholder name for applications that reach that stage like 
`New domain request`

![image](https://user-images.githubusercontent.com/52677065/205989540-779c02b1-e5e6-44ae-9f42-b418596a2080.png)

Error:
<h1 style="padding: 0px; margin: 0px; font-weight: normal; color: rgb(0, 0, 0); font-family: sans-serif; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">ValueError at /register/current_sites/</h1><pre class="exception_value" style="padding: 0px; margin: 10px 0px; font-size: 1.5em; white-space: pre-wrap; word-break: break-word; font-family: sans-serif; color: rgb(87, 87, 87); font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">You must specify a tld for </pre>

Request Method: | POST
-- | --
http://localhost:8080/register/current_sites/
4.1.3
ValueError
You must specify a tld for
/app/registrar/models/domain.py, line 136, in normalize
registrar.forms.application_wizard.ApplicationWizard
/usr/local/bin/python
3.10.8
['/app',  '/usr/local/lib/python310.zip',  '/usr/local/lib/python3.10',  '/usr/local/lib/python3.10/lib-dynload',  '/usr/local/lib/python3.10/site-packages']
Tue, 06 Dec 2022 18:14:06 +0000

Environment:


Request Method: POST
Request URL: http://localhost:8080/register/current_sites/

Django Version: 4.1.3
Python Version: 3.10.8
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'djangooidc',
 'auditlog',
 'widget_tweaks',
 'django_fsm',
 'registrar',
 'api',
 'nplusone.ext.django',
 'debug_toolbar']
Installed Middleware:
['debug_toolbar.middleware.DebugToolbarMiddleware',
 'allow_cidr.middleware.AllowCIDRMiddleware',
 'whitenoise.middleware.WhiteNoiseMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'csp.middleware.CSPMiddleware',
 'auditlog.middleware.AuditlogMiddleware',
 'nplusone.ext.django.NPlusOneMiddleware']



Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.10/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.10/contextlib.py", line 79, in inner
    return func(*args, **kwds)
  File "/usr/local/lib/python3.10/site-packages/django/views/generic/base.py", line 103, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/contrib/auth/mixins.py", line 73, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/formtools/wizard/views.py", line 244, in dispatch
    response = super().dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/views/generic/base.py", line 142, in dispatch
    return handler(request, *args, **kwargs)
  File "/app/registrar/forms/application_wizard.py", line 459, in post
    rendered = super().post(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/formtools/wizard/views.py", line 704, in post
    return super().post(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/formtools/wizard/views.py", line 296, in post
    self.storage.set_step_data(self.steps.current, self.process_step(form))
  File "/app/registrar/forms/application_wizard.py", line 441, in process_step
    self.forms_to_database(form=form)
  File "/app/registrar/forms/application_wizard.py", line 430, in forms_to_database
    form.to_database(application)
  File "/app/registrar/forms/application_wizard.py", line 123, in to_database
    normalized = Domain.normalize(self.cleaned_data["current_site"])
  File "/app/registrar/models/domain.py", line 136, in normalize
    raise ValueError("You must specify a tld for %s" % domain)

Exception Type: ValueError at /register/current_sites/
Exception Value: You must specify a tld for

SSPJ commented:
@igorkorenfeld Nice catch and thank you for testing this! I've added a flag to `normalize` to allow it to accept empty strings. I've also updated the in progress application table with your suggested default name.
